src/parser/SparqlParser.cpp

```diff
@@ -9,541 +9,426 @@
-                valuesStart++;
-              }
-              values._values.back()[i] = readTriplePart(inner, &valuesStart);
-              if (valuesStart >= inner.size() || valuesStart >= k) {
-                throw ParseException(
-                    "Expected " + std::to_string(values._variables.size()) +
-                    " values in every row of the VALUES block of " +
-                    inner.substr(valuesStatementStart, 256));
-              }
-            }
-          }
-          if (k == inner.size()) {
-            throw ParseException("A values block is not properly terminated: " +
-                                 inner.substr(valuesStatementStart, 256));
-          }
+      } else {
+        subject = lastSubject;
+        lastSubject.clear();
+      }
 
+      std::string predicate;
+      if (lastPredicate.empty()) {
+        if (_lexer.accept(SparqlToken::Type::VARIABLE)) {
+          predicate = _lexer.current().raw;
+        } else if (_lexer.accept(SparqlToken::Type::RDFLITERAL)) {
+          predicate = parseLiteral(_lexer.current().raw, true);
         } else {
-          // The values statement has only one variable
-          size_t var_start = k;
-          while (k < inner.size() && !std::isspace(inner[k]) &&
-                 inner[k] != '{') {
-            k++;
-          }
-          if (k == inner.size()) {
-            throw ParseException(
-                "Expected values after the variable definition of the VALUES "
-                "block, but reached the end of the input instead.");
-          }
-          values._variables.emplace_back(
-              inner.substr(var_start, k - var_start));
-          while (k < inner.size() && inner[k] != '{') {
-            k++;
-          }
-          // Skip the opening brace
-          k++;
-          if (k == inner.size()) {
-            throw ParseException(
-                "Expected values after the variable definition of the VALUES "
-                "block, but reached the end of the input instead.");
-          }
-          while (k < inner.size() && inner[k] != '}') {
-            while (k < inner.size() && std::isspace(inner[k])) {
-              ++k;
-            }
-            if (k == inner.size()) {
-              throw ParseException(
-                  "Expected a closing bracket but got no input for values "
-                  "statement " +
-                  inner.substr(valuesStatementStart, 256));
-            }
-            values._values.emplace_back();
-            values._values.back().push_back(
-                std::string(readTriplePart(inner, &k)));
-          }
-          if (k == inner.size()) {
-            throw ParseException("A values block is not properly terminated: " +
-                                 inner.substr(valuesStatementStart, 256));
-          }
+          // Assume the token is a predicate path. This will be verified
+          // separately later.
+          _lexer.expandNextUntilWhitespace();
```

Comment:
Are we sure there is actually whitespace after this? I assume this is covered by the `_re_string` being completely consumed, right?

Reply:
If the query is valid there has to be whitespace. Otherwise we will fail later on (either we reach the end of the input, or the property path is not valid, or there is a dot in place of a variable / iri).